src/basic/confidential-virt.c

```diff
@@ -0,0 +1,276 @@
+ * General-Purpose and System Instructions.
+ * Chapter: E4.1 - Maximum Extended Function Number and Vendor String
+ *  https://www.amd.com/system/files/TechDocs/24594.pdf
+ */
+#define CPUID_GET_HIGHEST_FUNCTION UINT32_C(0x80000000)
+
+/*
+ * AMD64 Architecture Programmer’s Manual Volume 3:
+ * General-Purpose and System Instructions.
+ * Chapter: E4.17 - Encrypted Memory Capabilities
+ *  https://www.amd.com/system/files/TechDocs/24594.pdf
+ */
+#define CPUID_AMD_GET_ENCRYPTED_MEMORY_CAPABILITIES UINT32_C(0x8000001f)
+
+/*
+ * AMD64 Architecture Programmer’s Manual Volume 3:
+ * General-Purpose and System Instructions.
+ * Chapter: 15.34.10 - SEV_STATUS MSR
+ * https://www.amd.com/system/files/TechDocs/24593.pdf
+ */
+#define MSR_AMD64_SEV UINT32_C(0xc0010131)
+
+/*
+ * Intel® TDX Module v1.5 Base Architecture Specification
+ * Chapter: 11.2
+ * https://www.intel.com/content/www/us/en/content-details/733575/intel-tdx-module-v1-5-base-architecture-specification.html
+ */
+
+#define CPUID_INTEL_TDX_ENUMERATION UINT32_C(0x21)
+
+/* Requirements for Implementing the Microsoft Hypervisor Interface
+ * https://learn.microsoft.com/en-us/virtualization/hyper-v-on-windows/tlfs/tlfs
+ */
+#define CPUID_HYPERV_VENDOR_AND_MAX_FUNCTIONS UINT32_C(0x40000000)
+
+#define CPUID_HYPERV_FEATURES UINT32_C(0x40000003)
+
+#define CPUID_HYPERV_ISOLATION_CONFIG UINT32_C(0x4000000C)
+
+#define CPUID_HYPERV_MIN UINT32_C(0x40000005)
+#define CPUID_HYPERV_MAX UINT32_C(0x4000ffff)
+
+#define CPUID_SIG_AMD       "AuthenticAMD"
+#define CPUID_SIG_INTEL     "GenuineIntel"
+#define CPUID_SIG_INTEL_TDX "IntelTDX    "
+#define CPUID_SIG_HYPERV    "Microsoft Hv"
+
+/* ecx bit 31: set => hyperpvisor, unset => bare metal */
+#define CPUID_FEATURE_HYPERVISOR (UINT32_C(1) << 31)
+
+/* Linux include/asm-generic/hyperv-tlfs.h */
+#define CPUID_HYPERV_CPU_MANAGEMENT (UINT32_C(1) << 12) /* root partition */
+#define CPUID_HYPERV_ISOLATION      (UINT32_C(1) << 22) /* confidential VM partition */
+
+#define CPUID_HYPERV_ISOLATION_TYPE_MASK UINT32_C(0xf)
+#define CPUID_HYPERV_ISOLATION_TYPE_SNP 2
+
+#if defined(__x86_64__)
+
+static void cpuid(uint32_t *eax, uint32_t *ebx, uint32_t *ecx, uint32_t *edx) {
+        log_debug("CPUID func %" PRIx32 " %" PRIx32, *eax, *ecx);
+        __cpuid_count(*eax, *ecx, *eax, *ebx, *ecx, *edx);
+        log_debug("CPUID result %" PRIx32 " %" PRIx32 " %" PRIx32 " %" PRIx32, *eax, *ebx, *ecx, *edx);
+}
+
+static uint32_t cpuid_leaf(uint32_t eax, char ret_sig[static 13], bool swapped) {
+        /* zero-init as some queries explicitly require subleaf == 0 */
+        uint32_t sig[3] = {};
+
+        if (swapped)
+                cpuid(&eax, &sig[0], &sig[2], &sig[1]);
+        else
+                cpuid(&eax, &sig[0], &sig[1], &sig[2]);
+        memcpy(ret_sig, sig, sizeof(sig));
+        ret_sig[12] = 0; /* \0-terminate the string to make string comparison possible */
+        log_debug("CPUID sig '%s'", ret_sig);
```

Comment:
This seems to sometimes print garbage to the console, making one of the tests fail as `expect` can't parse it as utf-8:

```
Detected virtualization systemd-nspawn.
CPUID func 1 0
CPUID result 50654 20000800 fffa3223 f8bfbff
CPUID is hypervisor: yes
CPUID func 0 0
CPUID result d 756e6547 6c65746e 49656e69
CPUID sig 'GenuineIntel'
CPUID func 80000000 0
CPUID result 80000008 756e6547 6c65746e 49656e69
CPUID func 21 0
CPUID result 2ff a88 a88 0
CPUID sig '€
'
Detected architecture x86-64.
Detected initialized system, this is not the first boot.
```

```
ERROR:test-shutdown:'utf-8' codec can't decode byte 0x88 in position 876: invalid start byte
```

Reply:
passing it through utf8_is_valid as a workaround

Reply:
I'm assuming that your VM is *not* running inside Intel TDX, and so the CPUID call we're doing is returning some nonsense value because this isn't TDX. This nonsense is then interpreted as a string when logging, and by chance this is resulting in bad utf-8.  Using `utf8_us_valid` looks sensible as a solution to this logging problem.